ddtrace/contrib/langchain/patch.py

```diff
@@ -0,0 +1,713 @@
+        "_api_key": os.getenv("DD_API_KEY"),
+    },
+)
+
+
+class _LangChainIntegration(BaseLLMIntegration):
+    _integration_name = "langchain"
+
+    def __init__(self, config, stats_url, site, api_key):
+        super().__init__(config, stats_url, site, api_key)
+
+    def _set_base_span_tags(self, span):
+        # type: (Span) -> None
+        return None
+
+    def _logs_tags(self, span):
+        # type: (Span) -> str
+        api_key = span.get_tag("langchain.request.api_key") or ""
+        tags = "env:%s,version:%s,langchain.request.provider:%s,langchain.request.model:%s,langchain.request.type:%s,langchain.request.api_key:%s" % (  # noqa: E501
+            (config.env or ""),
+            (config.version or ""),
+            (span.get_tag("langchain.request.provider") or ""),
+            (span.get_tag("langchain.request.model") or ""),
+            (span.get_tag("langchain.request.type") or ""),
+            api_key,
+        )
+        return tags
+
+    def _metrics_tags(self, span):
+        provider = span.get_tag("langchain.request.provider") or ""
+        api_key = span.get_tag("langchain.request.api_key") or ""
+        tags = [
+            "version:%s" % (config.version or ""),
+            "env:%s" % (config.env or ""),
+            "service:%s" % (span.service or ""),
+            "langchain.request.provider:%s" % provider,
+            "langchain.request.model:%s" % (span.get_tag("langchain.request.model") or ""),
+            "langchain.request.type:%s" % (span.get_tag("langchain.request.type") or ""),
+            "langchain.request.api_key:%s" % api_key,
+            "error:%d" % span.error,
+        ]
+        err_type = span.get_tag("error.type")
+        if err_type:
+            tags.append("error_type:%s" % err_type)
+        return tags
+
+    def record_usage(self, span, usage):
+        if not usage or not self._config.metrics_enabled:
+            return
+        for token_type in ("prompt", "completion", "total"):
+            num_tokens = usage.get("token_usage", {}).get(token_type + "_tokens")
+            if not num_tokens:
+                continue
+            self.metric(span, "dist", "tokens.%s" % token_type, num_tokens)
+        total_cost = span.get_metric("langchain.tokens.total_cost")
+        if total_cost:
+            self.metric(span, "incr", "tokens.total_cost", total_cost)
+
+
+def _extract_model_name(instance):
+    """Extract model name or ID from llm instance."""
+    for attr in ("model", "model_name", "model_id", "model_key", "repo_id"):
+        if hasattr(instance, attr):
+            return getattr(instance, attr, None)
+    return None
+
+
+def _extract_api_key(instance):
+    """
+    Extract and format LLM-provider API key from instance.
+    Note that langchain's LLM/ChatModel/Embeddings interfaces do not have a
+    standard attribute name for storing the provider-specific API key, so make a
+    best effort here by checking for attributes that end with `api_key/token`.
```

Comment:
```suggestion
    best effort here by checking for attributes that end with `api_key/api_token`.
```

Reply:
We make a best effort to extract keys, names and configurations from LLM models. Should we also add debug logging for when our best efforts fail. This could help us debug issues down the line and provide more insight on which models and versions of the api are not actually supported.

Reply:
I agree adding debug logging would be super helpful. This might add more detail to this PR which is already way too big to begin with though, so I'll add that in a future PR.